src/components/inputs/text/NameTokenSelector/NameTokenSelector.tsx

```diff
@@ -30,177 +44,512 @@ function NameTokenSelector({
+          ids.map(async (id) => {
+            const state =
+              await arweaveDataProvider.getContractState<PDNTContractJSON>(id);
+            return [id, state];
+          }),
+        );
+      if (!contracts) {
+        throw new Error('Unable to get details for Name Tokens');
+      }
+      const newTokens: {
+        [x: string]: Pick<
+          PDNTContractJSON,
+          'name' | 'ticker' | 'owner' | 'controller'
+        > & { names?: string[] };
+      } = contracts.reduce((result, contract) => {
+        const [id, state] = contract;
+        const { owner, controller, name, ticker } = state;
+        const associatedNames = getAssociatedNames(
+          id,
+          pdnsSourceContract.records,
+        );
+        return {
+          ...result,
+          [id.toString()]: {
+            owner,
+            controller,
+            name,
+            ticker,
+            names: associatedNames,
+          },
+        };
+      }, {});
+      setTokens(newTokens);
+      if (imports && newTokens) {
+        const details = newTokens[imports[0].toString()];
+        setSelectedToken({
+          name: details?.name,
+          ticker: details?.ticker,
+          id: imports[0].toString(),
+        });
+      }
+    } catch (error: any) {
+      eventEmitter.emit('error', error);
+    } finally {
+      setLoading(false);
+      setListPage(1);
+    }
+  }
+
+  function handleTokenSearch(query: string) {
     try {
       setSearching(true);
-      if (!id.length) {
+      setValidImport(undefined);
+      if (!query) {
         setSearchText('');
-        return;
+        setSelectedToken(undefined);
       }
+      setSearchText(query);
 
-      setSearchText(id);
-      const txid = new ArweaveTransactionID(id);
-      selectedTokenCallback(txid);
-      const state =
-        await arweaveDataProvider.getContractState<PDNTContractJSON>(txid);
-      const contract = new PDNTContract(state, txid);
-
-      if (!contract.isValid()) {
-        throw Error('Contract does not match required schema');
+      if (!tokens) {
+        throw new Error('No Name Tokens Found');
       }
-      setSelectedNameToken({
-        id,
-        name: contract.name,
-        ticker: contract.ticker,
-      });
-      setValidImport(true);
-      setSearchText('');
+      const filteredResults = Object.keys(tokens)
+        .map((id) => {
+          const { owner, controller, name, ticker } = tokens[id];
+          const queryResult = [owner, controller, name, ticker, id].map(
+            (term) =>
+              term
+                ? term
+                    .toString()
+                    .toLocaleLowerCase()
+                    .includes(query.toLocaleLowerCase())
+                : false,
+          );
+          if (queryResult.includes(true)) {
+            return { id: id, name: name ?? '', ticker: ticker ?? '' };
+          }
+        })
+        .filter((n) => n !== undefined);
+      if (!filteredResults) {
```

Comment:
Same here: this is a non-nullable array. Will always return false.

Reply:
Also, by applying the mapping first (which may return undefined) and then the filter, the code is inferring the type so it ends up having nullable items (even tho you filtered them out). Maybe applying the filter first, and then the mapping is the way to go.

Take a look to my chat with GPT - https://chat.openai.com/share/5ad748a1-fcab-4340-be7d-1a3edfa8b0f0